src/builder/QuarkBuilder.sol

```diff
@@ -99,6 +89,10 @@ contract QuarkBuilder {
             uint256 balanceOnDstChain =
                 Accounts.getBalanceOnChain(transferIntent.assetSymbol, transferIntent.chainId, chainAccountsList);
             uint256 amountLeftToBridge = transferIntent.amount - balanceOnDstChain;
+            // If the payment token is the transfer token and user opt for paying with the payment token, need to add max cost back to the amountLeftToBridge for target chain
+            if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, transferIntent.assetSymbol)) {
+                amountLeftToBridge += PaymentInfo.findMaxCost(payment, transferIntent.chainId);
+            }
```

Comment:
Since when we see `needsBridgedFunds()`, meaning the target chain is having less than the amount request on transfer. So here we check again and see if the transferring token is indeed payment token and opt for pay with token.  If it's true, then we need to add the maxCost to the balanceLeftToBridge as the target chain's balance will be all be used to transfer when computing the bridging logics in `balanceOnDstChain`